3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt

```diff
@@ -9,88 +9,65 @@ package com.adyen.checkout.adyen3ds2
 import com.adyen.checkout.core.exception.ComponentException
 import com.adyen.checkout.core.log.LogUtil
 import com.adyen.checkout.core.log.Logger
-import com.adyen.checkout.redirect.handler.RedirectHandler
-import com.adyen.threeds2.AuthenticationRequestParameters
-import com.adyen.threeds2.ChallengeStatusReceiver
-import com.adyen.threeds2.CompletionEvent
-import com.adyen.threeds2.ProtocolErrorEvent
-import com.adyen.threeds2.RuntimeErrorEvent
 import com.adyen.threeds2.ThreeDS2Service
-import com.adyen.threeds2.Transaction
 import com.adyen.threeds2.customization.UiCustomization
-import com.adyen.threeds2.exception.InvalidInputException
-import com.adyen.threeds2.exception.SDKAlreadyInitializedException
 import com.adyen.threeds2.exception.SDKNotInitializedException
-import com.adyen.threeds2.exception.SDKRuntimeException
-import com.adyen.threeds2.parameters.ChallengeParameters
-import com.adyen.threeds2.util.AdyenConfigParameters
-import kotlinx.coroutines.CoroutineExceptionHandler
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.launch
-import org.json.JSONException
-import org.json.JSONObject
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
 
-@Suppress("TooManyFunctions")
 class Adyen3DS2Component(
     savedStateHandle: SavedStateHandle,
     application: Application,
     configuration: Adyen3DS2Configuration,
-    private val submitFingerprintRepository: SubmitFingerprintRepository,
-    private val adyen3DS2Serializer: Adyen3DS2Serializer,
-    private val redirectHandler: RedirectHandler,
+    private val adyen3DS2Delegate: Adyen3DS2Delegate,
 ) : BaseActionComponent<Adyen3DS2Configuration>(savedStateHandle, application, configuration),
-    ChallengeStatusReceiver,
     IntentHandlingComponent {
 
-    private var mTransaction: Transaction? = null
-    private var mUiCustomization: UiCustomization? = null
-    private var authorizationToken: String?
-        get() {
-            return savedStateHandle[AUTHORIZATION_TOKEN_KEY]
-        }
-        set(value) {
-            savedStateHandle[AUTHORIZATION_TOKEN_KEY] = value
-        }
+    init {
+        adyen3DS2Delegate.detailsFlow
+            .onEach { notifyDetails(it) }
+            .launchIn(viewModelScope)
 
-    override fun onCleared() {
-        super.onCleared()
-        Logger.d(TAG, "onCleared")
-        if (mTransaction != null) {
-            // We don't save the mTransaction reference if the ViewModel gets cleared.
-            sGotDestroyedWhileChallenging = true
+        adyen3DS2Delegate.exceptionFlow
+            .onEach { notifyException(it) }
+            .launchIn(viewModelScope)
+
+        adyen3DS2Delegate.eventFlow
+            .onEach { onEvent(it) }
+            .launchIn(viewModelScope)
+    }
+
+    private fun onEvent(event: Adyen3DS2Event) {
+        when (event) {
+            Adyen3DS2Event.CleanUp3DS2 -> {
+                try {
+                    ThreeDS2Service.INSTANCE.cleanup(getApplication())
```

Comment:
How bad is it if we inject application into the delegate and clean up from there? Trying to reduce responsibilities of the components

Reply:
I don't like having any context injected into a delegate, not even in a ViewModel. I'd rather delegate it to a fragment/activity. But I suppose we can inject the application into a delegate, that should not cause any memory leaks. But I'm not sure if it will work in tests

Reply:
Yeah maybe let's not do it, we'll add some code in the components doesn't matter

Reply:
Ohh I did it already, it didn't cause any problems

Reply:
👍